File: src/scrapers/providers/manhwaindo.ts

import * as cheerio from 'cheerio';
import { MangaSource, type ScrapedManga, type ScraperProvider, type MangaDetail, type MangaChapter, type ChapterData } from '../types';

export class ManhwaIndoScraper implements ScraperProvider {
    name = MangaSource.MANHWAINDO;
    private readonly baseUrl = 'https://www.manhwaindo.my/';

    private formatIndonesianDate(dateStr: string): string {
        const months: { [key: string]: string } = {
            'Januari': 'January', 'Februari': 'February', 'Maret': 'March',
            'April': 'April', 'Mei': 'May', 'Juni': 'June',
            'Juli': 'July', 'Agustus': 'August', 'September': 'September',
            'Oktober': 'October', 'November': 'November', 'Desember': 'December',
            'Agust': 'August', 'Okt': 'October', 'Nov': 'November', 'Des': 'December'
        };
        return dateStr.replace(/\b(Januari|Februari|Maret|April|Mei|Juni|Juli|Agustus|September|Oktober|November|Desember|Agust|Okt|Nov|Des)\b/g, (match) => months[match]);
    }

    async scrapePopular(): Promise<ScrapedManga[]> {
        try {
            console.log(`Scraping ${this.name}...`);
            const response = await fetch(this.baseUrl);
            if (!response.ok) {
                throw new Error(`Failed to fetch ${this.baseUrl}: ${response.statusText}`);
            }
            const html = await response.text();
            // await Bun.write('debug_manhwaindo.html', html); // debugging
            const $ = cheerio.load(html);
            const mangaList: ScrapedManga[] = [];

            // Selectors based on debug HTML:
            // Container: .utao .uta
            $('.utao .uta').each((_, element) => {
                const titleElement = $(element).find('.luf h4');
                const imgElement = $(element).find('.imgu img');
                const latestChapterElement = $(element).find('.luf ul li:nth-child(1) a');
                const prevChapterElement = $(element).find('.luf ul li:nth-child(2) a');
                const linkElement = $(element).find('.luf a.series');

                const title = titleElement.text().trim();
                const link = linkElement.attr('href') || '';
                // Prefer data-src or data-original for lazy loaded images
                const image = imgElement.attr('data-src') || imgElement.attr('data-original') || imgElement.attr('src') || '';
                const chapter = latestChapterElement.text().trim();
                const previous_chapter = prevChapterElement.text().trim();

                if (title && link) {
                    mangaList.push({
                        title,
                        image,
                        source: this.name,
                        chapter,
                        previous_chapter,
                        link
                    });
                }
            });

            console.log(`Found ${mangaList.length} manga from ${this.name}`);
            return mangaList;
        } catch (error) {
            console.error(`Error scraping ${this.name}:`, error);
            return [];
        }
    }

    async search(query: string): Promise<ScrapedManga[]> {
        try {
            console.log(`Searching ${this.name} for "${query}"...`);
            const url = `${this.baseUrl}?s=${encodeURIComponent(query)}`;
            const response = await fetch(url, {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
                    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
                    'Accept-Language': 'en-US,en;q=0.5',
                    'Referer': this.baseUrl
                }
            });
            if (!response.ok) throw new Error(`Failed to fetch search: ${response.status}`);

            const html = await response.text();
            const $ = cheerio.load(html);
            const mangaList: ScrapedManga[] = [];

            // Similar selectors to popular/latest
            const items = $('.listupd .bs');
            console.log(`[ManhwaIndo] Found ${items.length} items in search DOM.`);

            items.each((_, element) => {
                let title = $(element).find('.tt a').text().trim();
                if (!title) title = $(element).find('.tt').text().trim(); // Fallback

                const link = $(element).find('a').attr('href') || '';
                let image = $(element).find('img').attr('data-src') ||
                    $(element).find('img').attr('data-original') ||
                    $(element).find('img').attr('src') || '';
                const chapter = $(element).find('.epxs').text().trim();
                const rating = parseFloat($(element).find('.numscore').text().trim()) || 0;

                if (title && link) {
                    mangaList.push({
                        title,
                        image,
                        source: this.name,
                        chapter,
                        rating,
                        link
                    });
                } else {
                    console.log('[ManhwaIndo] Skip item due to missing title/link', title, link);
                }
            });

            console.log(`Found ${mangaList.length} results from ${this.name}`);
            return mangaList;
        } catch (error) {
            console.error(`Error searching ${this.name}:`, error);
            return [];
        }
    }

    async scrapeDetail(link: string): Promise<MangaDetail | null> {
        try {
            console.log(`Scraping detail ${link}...`);
            const response = await fetch(link, {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
                }
            });
            if (!response.ok) {
                console.error(`Failed to fetch detail: ${response.status}`);
                return null;
            }
            const html = await response.text();
            const $ = cheerio.load(html);

            const title = $('h1').first().text().trim();
            const image = $('.series-thumb img').attr('src') || $('.thumb img').attr('src') || '';
            const synopsis = $('.series-synopsys').text().trim() || $('.entry-content').text().trim();
            const genres = $('.series-genres a, .genre-info a').map((_, el) => $(el).text().trim()).get();
            const status = 'Ongoing';
            const author = $('.series-infolist li:contains("Author") span').text().trim() || 'Unknown';
            const rating = parseFloat($('.series-rating').text().trim()) || 0;

            const chapters: MangaChapter[] = [];
            $('.series-chapterlist li, #chapterlist li').each((_, element) => {
                const linkEl = $(element).find('a');
                // Use .chapternum or fallback to text excluding .chapterdate
                let chapTitle = linkEl.find('.chapternum').text().trim() || linkEl.find('.chapter-name').text().trim();

                if (!chapTitle) {
                    // Fallback to text but try to remove date
                    const clone = linkEl.clone();
                    clone.find('.chapterdate, .chapter-date').remove();
                    chapTitle = clone.text().trim();
                }

                const chapLink = linkEl.attr('href') || '';
                const rawReleased = linkEl.find('.chapterdate').text().trim() || linkEl.find('.chapter-date').text().trim() || $(element).find('.chapter-date').text().trim();
                const released = this.formatIndonesianDate(rawReleased);

                if (chapTitle && chapLink) {
                    chapters.push({
                        title: chapTitle,
                        link: chapLink,
                        released
                    });
                }
            });

            return {
                title,
                image,
                synopsis,
                genres,
                author,
                status,
                rating,
                chapters
            };

        } catch (error) {
            console.error(`Error scraping detail from ${this.name}:`, error);
            return null;
        }
    }

    async scrapeChapter(link: string): Promise<ChapterData | null> {
        try {
            console.log(`Scraping chapter ${link}...`);
            const response = await fetch(link, {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
                }
            });
            if (!response.ok) {
                console.error(`Failed to fetch chapter: ${response.status}`);
                return null;
            }
            const html = await response.text();
            const $ = cheerio.load(html);

            const images: string[] = [];
            $('#readerarea img, .reading-content img').each((_, element) => {
                const dataSrc = $(element).attr('data-src');
                const src = $(element).attr('src');
                const validSrc = dataSrc || src;

                if (validSrc && !validSrc.startsWith('data:image')) {
                    images.push(validSrc.trim());
                } else if (validSrc) {
                    // Check if it's a real base64 image or just tiny placeholder
                    if (validSrc.length > 1000) {
                        images.push(validSrc.trim());
                    }
                }
            });

            // Navigation - Parse from Script Content (JSON variables)
            const scriptContent = $('script').map((_, el) => $(el).html()).get().join(' ');

            // Regex for unique structure seen in debug: "prevUrl":"...","nextUrl":"..."
            let next: string | undefined;
            let prev: string | undefined;

            const nextMatch = scriptContent.match(/"nextUrl"\s*:\s*"([^"]*)"/);
            const prevMatch = scriptContent.match(/"prevUrl"\s*:\s*"([^"]*)"/);

            if (nextMatch && nextMatch[1]) {
                const url = nextMatch[1].replace(/\\/g, ''); // Unescape slashes
                if (url && url.startsWith('http')) {
                    next = url;
                }
            }

            if (prevMatch && prevMatch[1]) {
                const url = prevMatch[1].replace(/\\/g, ''); // Unescape slashes
                if (url && url.startsWith('http')) {
                    prev = url;
                }
            }

            return { images, next, prev };
        } catch (error) {
            console.error(`Error scraping chapter from ${this.name}:`, error);
            return null;
        }
    }
}

File: internal/scraper/providers/manhwaindo.go
package providers

import (
	"context"
	"fmt"
	"regexp"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/gede-cahya/komida-backend/internal/scraper"
)

type ManhwaIndoScraper struct{}

func (s *ManhwaIndoScraper) Name() string { return "ManhwaIndo" }

func (s *ManhwaIndoScraper) reroute(link string) string {
	u, err := parseURL(link)
	if err != nil {
		return link
	}
	u.Scheme = "https"
	u.Host = "www.manhwaindo.my"
	return u.String()
}

func (s *ManhwaIndoScraper) formatDate(dateStr string) string {
	months := map[string]string{
		"Januari": "January", "Februari": "February", "Maret": "March",
		"April": "April", "Mei": "May", "Juni": "June",
		"Juli": "July", "Agustus": "August", "September": "September",
		"Oktober": "October", "November": "November", "Desember": "December",
		"Agust": "August", "Okt": "October", "Nov": "November", "Des": "December",
	}
	re := regexp.MustCompile(`\b(` + strings.Join(keysOf(months), "|") + `)\b`)
	return re.ReplaceAllStringFunc(dateStr, func(m string) string {
		if v, ok := months[m]; ok {
			return v
		}
		return m
	})
}

func (s *ManhwaIndoScraper) ScrapePopular(ctx context.Context) ([]scraper.ScrapedManga, error) {
	resp, err := fetchWithContext(ctx, "https://www.manhwaindo.my/", nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, err
	}
	var list []scraper.ScrapedManga
	doc.Find(".utao .uta").Each(func(_ int, el *goquery.Selection) {
		title := strings.TrimSpace(el.Find(".luf h4").Text())
		link, _ := el.Find(".luf a.series").Attr("href")
		imgEl := el.Find(".imgu img")
		image := coalesceAttr(imgEl, "data-src", "data-original", "src")
		chapter := strings.TrimSpace(el.Find(".luf ul li:nth-child(1) a").Text())
		prevChapter := strings.TrimSpace(el.Find(".luf ul li:nth-child(2) a").Text())
		if title != "" && link != "" {
			list = append(list, scraper.ScrapedManga{
				Title:           title,
				Image:           image,
				Source:          s.Name(),
				Chapter:         chapter,
				PreviousChapter: strPtr(prevChapter),
				Link:            link,
			})
		}
	})
	return list, nil
}

func (s *ManhwaIndoScraper) Search(ctx context.Context, query string) ([]scraper.ScrapedManga, error) {
	url := fmt.Sprintf("https://www.manhwaindo.my/?s=%s", urlEncode(query))
	headers := map[string]string{
		"User-Agent":      userAgent,
		"Accept":          "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
		"Accept-Language": "en-US,en;q=0.5",
		"Referer":         "https://www.manhwaindo.my/",
	}
	resp, err := fetchWithContext(ctx, url, headers)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, err
	}
	var list []scraper.ScrapedManga
	doc.Find(".listupd .bs").Each(func(_ int, el *goquery.Selection) {
		title := strings.TrimSpace(el.Find(".tt a").Text())
		if title == "" {
			title = strings.TrimSpace(el.Find(".tt").Text())
		}
		link, _ := el.Find("a").Attr("href")
		imgEl := el.Find("img")
		image := coalesceAttr(imgEl, "data-src", "data-original", "src")
		chapter := strings.TrimSpace(el.Find(".epxs").Text())
		if title != "" && link != "" {
			list = append(list, scraper.ScrapedManga{
				Title:   title,
				Image:   image,
				Source:  s.Name(),
				Chapter: chapter,
				Link:    link,
			})
		}
	})
	return list, nil
}

func (s *ManhwaIndoScraper) ScrapeDetail(ctx context.Context, link string) (*scraper.MangaDetail, error) {
	link = s.reroute(link)
	resp, err := fetchWithContext(ctx, link, defaultHeaders())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, err
	}
	title := strings.TrimSpace(doc.Find("h1").First().Text())
	imgEl := doc.Find(".series-thumb img")
	if imgEl.Length() == 0 {
		imgEl = doc.Find(".thumb img")
	}
	image := coalesceAttr(imgEl, "data-src", "data-original", "src")
	synopsis := strings.TrimSpace(doc.Find(".series-synopsys").Text())
	if synopsis == "" {
		synopsis = strings.TrimSpace(doc.Find(".entry-content").Text())
	}
	var genres []string
	doc.Find(".series-genres a, .genre-info a").Each(func(_ int, el *goquery.Selection) {
		genres = append(genres, strings.TrimSpace(el.Text()))
	})
	author := strings.TrimSpace(doc.Find(".series-infolist li:contains('Author') span").Text())
	if author == "" {
		author = "Unknown"
	}
	var chapters []scraper.MangaChapter
	doc.Find(".series-chapterlist li, #chapterlist li").Each(func(_ int, el *goquery.Selection) {
		linkEl := el.Find("a")
		chapTitle := strings.TrimSpace(linkEl.Find(".chapternum").Text())
		if chapTitle == "" {
			chapTitle = strings.TrimSpace(linkEl.Find(".chapter-name").Text())
		}
		if chapTitle == "" {
			clone := linkEl.Clone()
			clone.Find(".chapterdate, .chapter-date").Remove()
			chapTitle = strings.TrimSpace(clone.Text())
		}
		chapLink, _ := linkEl.Attr("href")
		rawDate := strings.TrimSpace(linkEl.Find(".chapterdate").Text())
		if rawDate == "" {
			rawDate = strings.TrimSpace(linkEl.Find(".chapter-date").Text())
		}
		if rawDate == "" {
			rawDate = strings.TrimSpace(el.Find(".chapter-date").Text())
		}
		if chapTitle != "" && chapLink != "" {
			chapters = append(chapters, scraper.MangaChapter{
				Title:    chapTitle,
				Link:     chapLink,
				Released: s.formatDate(rawDate),
			})
		}
	})
	return &scraper.MangaDetail{
		Title:    title,
		Image:    image,
		Synopsis: synopsis,
		Genres:   genres,
		Author:   author,
		Status:   "Ongoing",
		Chapters: chapters,
	}, nil
}

func (s *ManhwaIndoScraper) ScrapeChapter(ctx context.Context, link string) (*scraper.ChapterData, error) {
	link = s.reroute(link)
	resp, err := fetchWithContext(ctx, link, defaultHeaders())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, err
	}
	var images []string
	doc.Find("#readerarea img, .reading-content img").Each(func(_ int, el *goquery.Selection) {
		dataSrc, _ := el.Attr("data-src")
		src, _ := el.Attr("src")
		valid := dataSrc
		if valid == "" {
			valid = src
		}
		if valid != "" && !strings.HasPrefix(valid, "data:image") {
			images = append(images, strings.TrimSpace(valid))
		} else if valid != "" && len(valid) > 1000 {
			images = append(images, strings.TrimSpace(valid))
		}
	})
	var next, prev string
	scriptContent := ""
	doc.Find("script").Each(func(_ int, el *goquery.Selection) {
		scriptContent += " " + el.Text()
	})
	if m := regexp.MustCompile(`"nextUrl"\s*:\s*"([^"]*)"`).FindStringSubmatch(scriptContent); m != nil {
		url := strings.ReplaceAll(m[1], `\\`, "")
		if strings.HasPrefix(url, "http") {
			next = url
		}
	}
	if m := regexp.MustCompile(`"prevUrl"\s*:\s*"([^"]*)"`).FindStringSubmatch(scriptContent); m != nil {
		url := strings.ReplaceAll(m[1], `\\`, "")
		if strings.HasPrefix(url, "http") {
			prev = url
		}
	}
	return &scraper.ChapterData{Images: images, Next: next, Prev: prev}, nil
}

func (s *ManhwaIndoScraper) ScrapeGenres(ctx context.Context) ([]scraper.GenreItem, error) {
	return nil, nil // Not implemented in TypeScript version
}

func (s *ManhwaIndoScraper) ScrapeByGenre(ctx context.Context, genre string, page int) ([]scraper.ScrapedManga, error) {
	return nil, nil // Not implemented in TypeScript version
}
